// @google/gemini-cli-core/dist/src/tools/ls.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import fs from 'fs';
import path from 'path';
import { BaseTool, Icon } from './tools.js';
import { Type } from '@google/genai';
import { SchemaValidator } from '../utils/schemaValidator.js';
import { makeRelative, shortenPath } from '../utils/paths.js';
import { DEFAULT_FILE_FILTERING_OPTIONS } from '../config/config.js';
/**
 * Implementation of the LS tool logic
 */
export class LSTool extends BaseTool {
    config;
    static Name = 'list_directory';
    constructor(config) {
        super(LSTool.Name, 'ReadFolder', 'Lists the names of files and subdirectories directly within a specified directory path. Can optionally ignore entries matching provided glob patterns.', Icon.Folder, {
            properties: {
                path: {
                    description: 'The absolute path to the directory to list (must be absolute, not relative)',
                    type: Type.STRING,
                },
                ignore: {
                    description: 'List of glob patterns to ignore',
                    items: {
                        type: Type.STRING,
                    },
                    type: Type.ARRAY,
                },
                file_filtering_options: {
                    description: 'Optional: Whether to respect ignore patterns from .gitignore or .geminiignore',
                    type: Type.OBJECT,
                    properties: {
                        respect_git_ignore: {
                            description: 'Optional: Whether to respect .gitignore patterns when listing files. Only available in git repositories. Defaults to true.',
                            type: Type.BOOLEAN,
                        },
                        respect_gemini_ignore: {
                            description: 'Optional: Whether to respect .geminiignore patterns when listing files. Defaults to true.',
                            type: Type.BOOLEAN,
                        },
                    },
                },
            },
            required: ['path'],
            type: Type.OBJECT,
        });
        this.config = config;
    }
    /**
     * Validates the parameters for the tool
     * @param params Parameters to validate
     * @returns An error message string if invalid, null otherwise
     */
    validateToolParams(params) {
        const errors = SchemaValidator.validate(this.schema.parameters, params);
        if (errors) {
            return errors;
        }
        if (!path.isAbsolute(params.path)) {
            return `Path must be absolute: ${params.path}`;
        }
        const workspaceContext = this.config.getWorkspaceContext();
        if (!workspaceContext.isPathWithinWorkspace(params.path)) {
            const directories = workspaceContext.getDirectories();
            return `Path must be within one of the workspace directories: ${directories.join(', ')}`;
        }
        return null;
    }
    /**
     * Checks if a filename matches any of the ignore patterns
     * @param filename Filename to check
     * @param patterns Array of glob patterns to check against
     * @returns True if the filename should be ignored
     */
    shouldIgnore(filename, patterns) {
        if (!patterns || patterns.length === 0) {
            return false;
        }
        for (const pattern of patterns) {
            // Convert glob pattern to RegExp
            const regexPattern = pattern
                .replace(/[.+^${}()|[\]\\]/g, '\\$&')
                .replace(/\*/g, '.*')
                .replace(/\?/g, '.');
            const regex = new RegExp(`^${regexPattern}$`);
            if (regex.test(filename)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Gets a description of the file reading operation
     * @param params Parameters for the file reading
     * @returns A string describing the file being read
     */
    getDescription(params) {
        const relativePath = makeRelative(params.path, this.config.getTargetDir());
        return shortenPath(relativePath);
    }
    // Helper for consistent error formatting
    errorResult(llmContent, returnDisplay) {
        return {
            llmContent,
            // Keep returnDisplay simpler in core logic
            returnDisplay: `Error: ${returnDisplay}`,
        };
    }
    /**
     * Executes the LS operation with the given parameters
     * @param params Parameters for the LS operation
     * @returns Result of the LS operation
     */
    async execute(params, _signal) {
        const validationError = this.validateToolParams(params);
        if (validationError) {
            return this.errorResult(`Error: Invalid parameters provided. Reason: ${validationError}`, `Failed to execute tool.`);
        }
        try {
            const stats = fs.statSync(params.path);
            if (!stats) {
                // fs.statSync throws on non-existence, so this check might be redundant
                // but keeping for clarity. Error message adjusted.
                return this.errorResult(`Error: Directory not found or inaccessible: ${params.path}`, `Directory not found or inaccessible.`);
            }
            if (!stats.isDirectory()) {
                return this.errorResult(`Error: Path is not a directory: ${params.path}`, `Path is not a directory.`);
            }
            const files = fs.readdirSync(params.path);
            const defaultFileIgnores = this.config.getFileFilteringOptions() ?? DEFAULT_FILE_FILTERING_OPTIONS;
            const fileFilteringOptions = {
                respectGitIgnore: params.file_filtering_options?.respect_git_ignore ??
                    defaultFileIgnores.respectGitIgnore,
                respectGeminiIgnore: params.file_filtering_options?.respect_gemini_ignore ??
                    defaultFileIgnores.respectGeminiIgnore,
            };
            // Get centralized file discovery service
            const fileDiscovery = this.config.getFileService();
            const entries = [];
            let gitIgnoredCount = 0;
            let geminiIgnoredCount = 0;
            if (files.length === 0) {
                // Changed error message to be more neutral for LLM
                return {
                    llmContent: `Directory ${params.path} is empty.`,
                    returnDisplay: `Directory is empty.`,
                };
            }
            for (const file of files) {
                if (this.shouldIgnore(file, params.ignore)) {
                    continue;
                }
                const fullPath = path.join(params.path, file);
                const relativePath = path.relative(this.config.getTargetDir(), fullPath);
                // Check if this file should be ignored based on git or gemini ignore rules
                if (fileFilteringOptions.respectGitIgnore &&
                    fileDiscovery.shouldGitIgnoreFile(relativePath)) {
                    gitIgnoredCount++;
                    continue;
                }
                if (fileFilteringOptions.respectGeminiIgnore &&
                    fileDiscovery.shouldGeminiIgnoreFile(relativePath)) {
                    geminiIgnoredCount++;
                    continue;
                }
                try {
                    const stats = fs.statSync(fullPath);
                    const isDir = stats.isDirectory();
                    entries.push({
                        name: file,
                        path: fullPath,
                        isDirectory: isDir,
                        size: isDir ? 0 : stats.size,
                        modifiedTime: stats.mtime,
                    });
                }
                catch (error) {
                    // Log error internally but don't fail the whole listing
                    console.error(`Error accessing ${fullPath}: ${error}`);
                }
            }
            // Sort entries (directories first, then alphabetically)
            entries.sort((a, b) => {
                if (a.isDirectory && !b.isDirectory)
                    return -1;
                if (!a.isDirectory && b.isDirectory)
                    return 1;
                return a.name.localeCompare(b.name);
            });
            // Create formatted content for LLM
            const directoryContent = entries
                .map((entry) => `${entry.isDirectory ? '[DIR] ' : ''}${entry.name}`)
                .join('\n');
            let resultMessage = `Directory listing for ${params.path}:\n${directoryContent}`;
            const ignoredMessages = [];
            if (gitIgnoredCount > 0) {
                ignoredMessages.push(`${gitIgnoredCount} git-ignored`);
            }
            if (geminiIgnoredCount > 0) {
                ignoredMessages.push(`${geminiIgnoredCount} gemini-ignored`);
            }
            if (ignoredMessages.length > 0) {
                resultMessage += `\n\n(${ignoredMessages.join(', ')})`;
            }
            let displayMessage = `Listed ${entries.length} item(s).`;
            if (ignoredMessages.length > 0) {
                displayMessage += ` (${ignoredMessages.join(', ')})`;
            }
            return {
                llmContent: resultMessage,
                returnDisplay: displayMessage,
            };
        }
        catch (error) {
            const errorMsg = `Error listing directory: ${error instanceof Error ? error.message : String(error)}`;
            return this.errorResult(errorMsg, 'Failed to list directory.');
        }
    }
}
